File: Archived_Code_20200906_1604/Interview_Preparation_20201007_1037/InterviewBit/Programming/LinkedLists/remove-nth-node-from-list-end.cpp
#include <bits/stdc++.h>

struct ListNode {
    int val;
    struct ListNode *next;
    explicit ListNode(int val) {
        this->val = val;
        this->next = nullptr;
    }
};

void print(ListNode *head) {
    ListNode *curr = head;
    while (curr) {
        std::cout << curr->val << " ";
        curr = curr->next;
    }
    std::cout << "\n";
}

int listLength(ListNode *head) {
    int count = 0;
    ListNode *curr = head;
    while (curr) {
        count++;
        curr = curr->next;
    }
    return count;
}

ListNode* removeNthFromEnd(ListNode* A, int B) {
    if (!A) return A;
    int N = listLength(A);
    if (B >= N) {
        ListNode *tmp = A;
        A = A->next;
        delete tmp;
        return A;
    } 
    ListNode *curr = A, *prev = nullptr;
    int count = 1;
    while (count < (N - B + 1)) {
        prev = curr;
        curr = curr->next;
        count++;
    }
    if (curr->next == nullptr) prev->next = nullptr;
    else {
        prev->next = curr->next;
    }
    if (!curr) delete curr;
    return A;
}

int main(int argc, char const *argv[])
{
    ListNode *head1 = new ListNode(1);
    head1->next = new ListNode(2);
    head1->next->next = new ListNode(3);
    head1->next->next->next = new ListNode(4);
    head1->next->next->next->next = new ListNode(5);
    print(head1);
    
    print(removeNthFromEnd(head1, 5));
    return 0;
}

File: 20210816-0947/Archived_Code_20200906_1604/Backup/Code_25032020_Backup_Do_Not_Delete/src/com/interviewbit/LinkedLists/RemoveNthNodeFromListEnd.java
package com.interviewbit.LinkedLists;

public class RemoveNthNodeFromListEnd {

    static class ListNode {
        int val;
        ListNode next;
        ListNode(int value) {
            val = value;
            next = null;
        }
    }

    public static int getListLength(final ListNode head) {
        int length = 0;
        ListNode curr = head;
        while (curr != null) {
            length++;
            curr = curr.next;
        }
        return length;
    }

    public static void printList(final ListNode head) {
        ListNode curr = head;
        while (curr != null) {
            System.out.print(curr.val + " ");
            curr = curr.next;
        }
        System.out.println();
    }

    public static ListNode removeNthFromEnd(ListNode head, int n) {
        if (head == null) return null;
        int length = getListLength(head);
        if (n >= length) {
            head = head.next;
            return head;
        }
        ListNode curr = head, prev = null;
        int diff = length - n + 1;
        while (--diff > 0) {
            prev = curr;
            curr = curr.next;
        }
        if (curr.next != null) prev.next = curr.next;
        else prev.next = null;
        return head;
    }

    public static void main(String[] args) {
        ListNode head = new ListNode(1);
        head.next = new ListNode(2);
        head.next.next = new ListNode(3);
        head.next.next.next = new ListNode(4);
        head.next.next.next.next = new ListNode(5);
        printList(head);
        printList(removeNthFromEnd(head, 5));
    }
}
